extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/output/RetriableS3OutputStream.java

```diff
@@ -0,0 +1,432 @@
+
+  private static final Logger LOG = new Logger(RetriableS3OutputStream.class);
+  private static final Joiner JOINER = Joiner.on("/").skipNulls();
+  private static final int S3_MULTIPART_UPLOAD_MAX_NUM_PARTS = 10_000;
+
+  private final S3OutputConfig config;
+  private final ServerSideEncryptingAmazonS3 s3;
+  private final String s3Key;
+  private final String uploadId;
+  private final File chunkStorePath;
+  private final long chunkSize;
+
+  private final List<PartETag> pushResults = new ArrayList<>();
+  private final byte[] singularBuffer = new byte[1];
+
+  // metric
+  private final Stopwatch pushStopwatch;
+
+  private Chunk currentChunk;
+  private int nextChunkId = 1; // multipart upload requires partNumber to be in the range between 1 and 10000
+  private int numChunksPushed;
+  /**
+   * Total size of all chunks. This size is updated whenever the chunk is ready for push,
+   * not when {@link #write(byte[], int, int)} is called. This is because
+   * it will be hard to know the increase of chunk size in write() when the chunk is compressed.
+   */
+  private long resultsSize;
+
+  /**
+   * A flag indicating whether there was an upload error.
+   * This flag is tested in {@link #close()} to determine whether it needs to upload the current chunk or not.
+   */
+  private boolean error;
+  private boolean closed;
+
+  public RetriableS3OutputStream(
+      S3OutputConfig config,
+      ServerSideEncryptingAmazonS3 s3,
+      String s3Key
+  ) throws IOException
+  {
+
+    this(config, s3, s3Key, true);
+  }
+
+  @VisibleForTesting
+  protected RetriableS3OutputStream(
+      S3OutputConfig config,
+      ServerSideEncryptingAmazonS3 s3,
+      String s3Key,
+      boolean chunkValidation
+  ) throws IOException
+  {
+    this.config = config;
+    this.s3 = s3;
+    this.s3Key = s3Key;
+
+    final InitiateMultipartUploadResult result = s3.initiateMultipartUpload(
+        new InitiateMultipartUploadRequest(config.getBucket(), s3Key)
+    );
+    this.uploadId = result.getUploadId();
+    this.chunkStorePath = new File(config.getTempDir(), uploadId + UUID.randomUUID());
+    FileUtils.mkdirp(this.chunkStorePath);
+    this.chunkSize = config.getChunkSize() == null ? computeChunkSize(config) : config.getChunkSize();
+    if (chunkValidation) {
+      validateChunkSize(config.getMaxResultsSize(), chunkSize);
+    }
+    this.pushStopwatch = Stopwatch.createUnstarted();
+    this.pushStopwatch.reset();
+
+    this.currentChunk = new Chunk(nextChunkId, new File(chunkStorePath, String.valueOf(nextChunkId++)));
+  }
+
+  private static long computeChunkSize(S3OutputConfig config)
+  {
+    return computeMinChunkSize(config.getMaxResultsSize());
+  }
+
+  private static void validateChunkSize(long maxResultsSize, long chunkSize)
+  {
+    if (computeMinChunkSize(maxResultsSize) > chunkSize) {
+      throw new IAE(
+          "chunkSize[%s] is too small for maxResultsSize[%s]. chunkSize should be at least [%s]",
```

Comment:
`%s` for a number. `%d` is preferred. Please add spaces before brackets, though brackets are not really needed for numbers.

Maybe we can improve the message. "(What?) selected an S3 upload chunk size of %d bytes, but the configuration requires at least %d bytes, due to a configured maxResultSize of %d bytes."

But, taking a step back, I'm not sure that I, as a user, would have any idea what to do. Why should I care about chunk size? How does the maximum file size relate to chunk size? More to the point, why didn't Druid do the math itself to pick the chunk size? Even more, if the chunks are too small, why can't Druid just concatenate several of them to get its preferred size?

There is probably a reason. Can we maybe add some Javadoc to explain? Or, if this is copy/pasted from existing code, and that code is documented, perhaps point to the docs.

Reply:
Yeah the user of this class should not be worried about the chunksize. 
Move the logic to s3OutputConfig for validation there.